servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,257 @@
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.Publisher;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.MessageToByteEncoder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import static io.netty.util.internal.PlatformDependent.throwException;
+import static io.servicetalk.buffer.api.EmptyBuffer.EMPTY_BUFFER;
+import static io.servicetalk.buffer.api.ReadOnlyBufferAllocators.DEFAULT_RO_ALLOCATOR;
+import static io.servicetalk.buffer.netty.BufferUtils.newBufferFrom;
+import static io.servicetalk.buffer.netty.BufferUtils.toByteBuf;
+import static io.servicetalk.concurrent.api.Single.succeeded;
+import static java.util.Objects.requireNonNull;
+
+final class NettyChannelContentCodec extends AbstractContentCodec {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyChannelContentCodec.class);
+    private static final Buffer END_OF_STREAM = DEFAULT_RO_ALLOCATOR.fromAscii(" ");
+
+    private final Supplier<MessageToByteEncoder<ByteBuf>> encoderFn;
+    private final Supplier<ByteToMessageDecoder> decoderFn;
+
+    NettyChannelContentCodec(final CharSequence name,
+                             final Supplier<MessageToByteEncoder<ByteBuf>> encoderFn,
+                             final Supplier<ByteToMessageDecoder> decoderFn) {
+        super(name);
+        this.encoderFn = encoderFn;
+        this.decoderFn = decoderFn;
+    }
+
+    @Override
+    public Buffer encode(final Buffer src, final int offset, final int length,
+                         final BufferAllocator allocator) {
+        requireNonNull(src);
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0)");
+        }
+
+        final MessageToByteEncoder<ByteBuf> encoder = encoderFn.get();
+        final EmbeddedChannel channel = new EmbeddedChannel(encoder);
+
+        try {
+            ByteBuf origin = toByteBufRetained(src, offset, length);
+            channel.writeOutbound(origin);
+
+            // May produce footer
+            channel.close();
+
+            CompositeBuffer compositeBuffer = allocator.newCompositeBuffer();
+            ByteBuf chunk;
+            while ((chunk = channel.readOutbound()) != null) {
+                compositeBuffer.addBuffer(newBufferFromRetained(chunk));
+            }
+
+            return compositeBuffer;
+        } catch (Exception e) {
+            LOGGER.error("Error while encoding with {}", name(), e);
+            throwException(e);
+            return EMPTY_BUFFER;
+        } finally {
+            safeClose(channel);
```

Comment:
Do we need `safeClose` in `finally` if we already closed inside `try`? Can we do this only inside `catch` block? If you will move it to the `catch` block only, do we need `channel.checkException()` after closing the channel inside `try` block? Or maybe just invoke `finish` there?